katdal/chunkstore_s3.py

```diff
@@ -398,35 +438,159 @@ class S3ChunkStore(ChunkStore):
         If S3 server interaction failed (it's down, no authentication, etc)
     """
 
-    def __init__(self, url, timeout=300, token=None, credentials=None,
-                 public_read=False, expiry_days=0, **kwargs):
+    def __init__(self, url, timeout=(30, 300), retries=2, token=None,
+                 credentials=None, public_read=False, expiry_days=0, **kwargs):
         error_map = {requests.exceptions.RequestException: StoreUnavailable,
                      defusedxml.ElementTree.ParseError: StoreUnavailable}
         super(S3ChunkStore, self).__init__(error_map)
         auth = _auth_factory(url, token, credentials)
-
-        def session_factory(timeout=timeout, retries=2):
+        if not isinstance(retries, Retry):
+            try:
+                connect_retries, read_retries = retries
+            except TypeError:
+                connect_retries = read_retries = retries
+            # The backoff factor of 10 provides 5 minutes worth of retries
+            # when the Ceph gateway is strained; with 5 retries you get
+            # (0 + 2 + 4 + 8 + 16) * 10 = 300 seconds on top of read timeouts.
+            retries = Retry(connect=connect_retries, read=read_retries,
+                            status=5, backoff_factor=10.,
+                            status_forcelist=_DEFAULT_SERVER_GLITCHES)
+
+        def session_factory():
             session = _CacheSettingsSession(url)
             session.auth = auth
-            adapter = _TimeoutHTTPAdapter(max_retries=retries, timeout=timeout)
+            # Don't let requests do status retries as we'll be doing it ourselves
+            max_retries = retries.new(status=0, raise_on_status=False)
+            adapter = requests.adapters.HTTPAdapter(max_retries=max_retries)
```

Comment:
I only take over the "status" retries logic, leaving the rest for `requests`.

It could still happen that the total retries are breached if e.g. the retries goes `read, status, status, read, status, status, read, ...`. I've tried to track the total retries externally but have given up for now (and it should not happen that often anyway).